Fix crashes from hardening _WKRemoteObjectRegistry decoding

achristensen07 opened the pull request:
#### 80c45a1e6ff21d51c26acad87b0a02f35357e202
<pre>
Fix crashes from hardening _WKRemoteObjectRegistry decoding
<a href="https://bugs.webkit.org/show_bug.cgi?id=262983">https://bugs.webkit.org/show_bug.cgi?id=262983</a>
rdar://113527046

Reviewed by David Kilzer.

In April 2023, I made _WKRemoteObjectRegistry deserialization stricter by not allowing subclasses
of the specified class to be deserialized.  To make this transition as smooth as possible, I added
a set of common and safe always-allowed subclasses including NSMutableString and several others.
Telemetry indicates this increased the crash rate and this is to bring that crash rate back down.

After analyzing 100% of the recent crash reports with useful data, I found 4 things that will help:
1. Allowing a class named &quot;NSDecimalNumberPlaceholder&quot; which is a subclass of NSDecimalNumber that
   some internal frameworks apparently use and give to Safari in a path that serializes it.  Allow this.
2. Some crash logs indicate that NSDate objects are failing to decode because they are not in the
   set of allowed classes.  There are some JS to ObjC object converters (such as the one used in
   WKWebView.callAsyncJavaScript) that can produce an NSDate, and this is ok and safe.  Allow this too.
3. There are logs that indicate that sometimes a class is being sent from one process to another,
   and the receiving process has not loaded the dylib containing the ObjC class so the ObjC runtime
   can&apos;t find it.  Add telemetry to get this class name for future diagnosis.
4. There are logs that indicate that sometimes a class is being sent that does not conform to
   NSSecureCoding according to NSCoder.  Also add telemetry to get this class name for future diagnosis,
   but in this case I needed to add a @try/@catch because validateClassSupportsSecureCoding either
   returns YES or throws an ObjC exception, so to get the class name to CRASH_WITH_INFO I need to catch.

* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(alwaysAllowedClasses):
(validateClass):
(decodeObject):

Canonical link: <a href="https://commits.webkit.org/269185@main">https://commits.webkit.org/269185@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b82991d43629aea69a8f8748a3bfced38598b78

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20200 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26277 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22346 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18903 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24545 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18803 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26038 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19974 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23903 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20434 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19777 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2711 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20373 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
```diff
@@ -799,6 +799,8 @@ static id decodeObjectFromObjectStream(WKRemoteObjectDecoder *decoder, const Has
         (__bridge CFTypeRef)NSError.class,
         (__bridge CFTypeRef)NSURLError.class,
+        (__bridge CFTypeRef)NSDate.class,
+        (__bridge CFTypeRef)NSClassFromString(@"NSDecimalNumberPlaceholder"),
```
Do you need to add `NSDecimalNumber` explicitly as well?

achristensen07 replied:
Apparently not, but there's no reason not to add it while allowing its subclass.

ddkilzer commented on Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
```diff
@@ -831,8 +833,15 @@ static void validateClass(WKRemoteObjectDecoder *decoder, Class objectClass)
+        NSString *message = [NSString stringWithFormat:@"%@,%@", NSStringFromClass(objectClass), exception];
+        std::array<uint64_t, 6> values { 0, 0, 0, 0, 0, 0 };
+        strncpy(reinterpret_cast<char*>(values.data()), message.UTF8String, sizeof(values));
+        CRASH_WITH_INFO(values[0], values[1], values[2], values[3], values[4], values[5]);
```
For the future, it might be nice to extract this into an inline method or extract it into its own `CRASH()` macro that takes a C-string (since this is the second and third copy of this code, IIRC).  Not required to fix for this PR.

ddkilzer reviewed (approved):
r=me

achristensen07 committed:
Fix crashes from hardening _WKRemoteObjectRegistry decoding
https://bugs.webkit.org/show_bug.cgi?id=262983
rdar://113527046

Reviewed by David Kilzer.

In April 2023, I made _WKRemoteObjectRegistry deserialization stricter by not allowing subclasses
of the specified class to be deserialized.  To make this transition as smooth as possible, I added
a set of common and safe always-allowed subclasses including NSMutableString and several others.
Telemetry indicates this increased the crash rate and this is to bring that crash rate back down.

After analyzing 100% of the recent crash reports with useful data, I found 4 things that will help:
1. Allowing a class named "NSDecimalNumberPlaceholder" which is a subclass of NSDecimalNumber that
   some internal frameworks apparently use and give to Safari in a path that serializes it.  Allow this.
2. Some crash logs indicate that NSDate objects are failing to decode because they are not in the
   set of allowed classes.  There are some JS to ObjC object converters (such as the one used in
   WKWebView.callAsyncJavaScript) that can produce an NSDate, and this is ok and safe.  Allow this too.
3. There are logs that indicate that sometimes a class is being sent from one process to another,
   and the receiving process has not loaded the dylib containing the ObjC class so the ObjC runtime
   can't find it.  Add telemetry to get this class name for future diagnosis.
4. There are logs that indicate that sometimes a class is being sent that does not conform to
   NSSecureCoding according to NSCoder.  Also add telemetry to get this class name for future diagnosis,
   but in this case I needed to add a @try/@catch because validateClassSupportsSecureCoding either
   returns YES or throws an ObjC exception, so to get the class name to CRASH_WITH_INFO I need to catch.

* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(alwaysAllowedClasses):
(validateClass):
(decodeObject):

Canonical link: https://commits.webkit.org/269185@main
